apis/controlplane/v1alpha3/nestedcontrolplane_types.go

```diff
@@ -0,0 +1,107 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha3
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// NestedControlPlaneSpec defines the desired state of NestedControlPlane
+type NestedControlPlaneSpec struct {
+	// ClusterNamespace allows you to specify the namespace where the control
+	// plane components are run, if you don't specify one then the namespace
+	// will be in the format of {namespace}-{vcp.uid}. This also makes up the
+	// prefixes used for all Namespaces in the cluster.
+	// e.g. {namespace}-{vcp.uid}-{vc-namespace}
+	// +optional
+	ClusterNamespace string `json:"clusterNamespace,omitempty"`
```

Comment:
What's the `vcp.uid`? We should make these doc comments as clear as possible for the 'layman' user and avoid ambiguous terms if possible 😄

> This also makes up the prefixes used for all Namespaces in the cluster. e.g. {namespace}-{vcp.uid}-{vc-namespace}

This seems okay for now. In future though it may be worth having some kind of `namespacePrefix` field to explicitly set this to some other value. It may be worth noting here that whilst it is true today, it should not be relied upon? If so, can we expose the chosen prefix through a status field to make it easier for other consumers of this API to make use of/build upon?

Reply:
Also, I think there needs to be a length restriction on this field. We should also call out how the length of this field influences the restrictions on namespace names within nested clusters too.

Reply:
+1 for the `namespacePrefix`

Reply:
We need to call out this is typically defaulted and that it needs to have a max length validation and warning about it's usage for conflicting namespace prefixes, also rename to `NamespacePrefix`